### PR TITLE
Change unfocus to unfocus the entire chain, not just the primary focus

### LINKED ISSUE #7247: Dismissing keyboard programatically
It's not clear how to accomplish this.

### LINKED ISSUE #31784: The method 'detach' isn't defined for the class 'FocusScopeNode'
After a flutter upgrade and a packages upgrade I'm having this error:

**The method 'detach' isn't defined for the class 'FocusScopeNode'** on:

`FocusScope.of(context).detach();
`

> [✓] Flutter (Channel master, v1.5.9-pre.46, on Mac OS X 10.14.4 18E226, locale it-IT)
>     • Flutter version 1.5.9-pre.46 at /Users/miottoan/flutter
>     • Framework revision e62b06a3ea (27 minutes ago), 2019-04-29 05:58:18 -0400
>     • Engine revision 48a5ce87ae
>     • Dart version 2.3.0 (build 2.3.0-dev.0.3 c46deebfb6)
> 
> [✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
>     • Android SDK at /Users/miottoan/Library/Android/sdk
>     • Android NDK location not configured (optional; useful for native profiling support)
>     • Platform android-28, build-tools 28.0.3
>     • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
>     • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
>     • All Android licenses accepted.
> 
> [✓] iOS toolchain - develop for iOS devices (Xcode 10.2.1)
>     • Xcode at /Applications/Xcode.app/Contents/Developer
>     • Xcode 10.2.1, Build version 10E1001
>     • ios-deploy 1.9.4
>     • CocoaPods version 1.6.0
> 
> [✓] Android Studio (version 3.4)
>     • Android Studio at /Applications/Android Studio.app/Contents
>     • Flutter plugin version 35.0.1
>     • Dart plugin version 183.5901
>     • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
> 
> [✓] VS Code (version 1.33.1)
>     • VS Code at /Applications/Visual Studio Code.app/Contents
>     • Flutter extension version 2.26.0
> 
> [✓] Connected device (1 available)
>     • iPhone Xʀ • 28F1BE2E-AE5F-4CAF-A1D8-0FBD47BA45FA • ios • com.apple.CoreSimulator.SimRuntime.iOS-12-2 (simulator)

### LINKED ISSUE #31827: Unable get focus of TextFormField - keyboard disappears
## Steps to Reproduce

Run this sample, click the body button "Tab 1" - the route will appear modally, you will see the keyboard attempt to show - then disappear.

You are then unable to get focus of any textfield, this never used to be an issue in my app until I upgraded flutter yesterday.

```
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CupertinoTabScaffold(
      tabBar: CupertinoTabBar(
        items: [
          BottomNavigationBarItem(icon: Icon(Icons.message)),
          BottomNavigationBarItem(icon: Icon(Icons.block)),
        ],
      ),
      tabBuilder: (BuildContext context, int index) {
        switch (index) {
          case 0:
            return CupertinoTabView(
              builder: (context) => Page1(
                    title: "Tab 1",
                  ),
            );
            break;
          case 1:
            return CupertinoTabView(
              builder: (context) => Page1(
                    title: "Tab 2",
                  ),
            );
            break;
        }
      },
    );
  }
}

class Page1 extends StatelessWidget {
  final String title;

  const Page1({Key key, this.title}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: Center(
        child: FlatButton(
          child: Text(title),
          onPressed: () {
            Navigator.of(context, rootNavigator: true).push(
              MaterialPageRoute(
                fullscreenDialog: true,
                builder: (context) => CreateEditProductWizardPage(),
              ),
            );
          },
        ),
      ),
    );
  }
}

class CreateEditProductWizardPage extends StatefulWidget {
  final bool editing;

  const CreateEditProductWizardPage({Key key, this.editing = false})
      : super(key: key);

  @override
  _CreateEditProductWizardPageState createState() =>
      _CreateEditProductWizardPageState();
}

class _CreateEditProductWizardPageState
    extends State<CreateEditProductWizardPage> {
  static final _formKey = GlobalKey<FormState>();

  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Form(
        key: _formKey,
        child: ListView(
          children: [
            TextFormField(
              autofocus: true,
              maxLines: 1,
            ),
            Divider(
              height: 0,
            ),
            TextFormField(
              maxLines: 2,
            ),
            Divider(
              height: 0,
            ),
            TextFormField(
              textInputAction: TextInputAction.done,
              keyboardType: TextInputType.number,
            ),
            Divider(
              height: 0,
            ),
            TextFormField(
              textInputAction: TextInputAction.done,
              keyboardType: TextInputType.number,
            ),
          ],
        ),
      ),
    );
  }
}
```

## Logs

Nothing produced in the logs

```
[✓] Flutter (Channel master, v1.5.9-pre.66, on Mac OS X 10.14.4 18E226, locale en-ZA)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 10.2.1)
[✓] Android Studio (version 3.3)
[✓] IntelliJ IDEA Ultimate Edition (version 2018.3.2)
[✓] Connected device (1 available)

```

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

In https://github.com/flutter/flutter/pull/31614, I added an `unfocus()` to FocusNodes to allow giving up of focus, but it only worked on the primary focus.  This PR changes that so that it will unfocus the entire chain, not just the primary focus. Now, if you call `unfocus()` on a `FocusNode` or `FocusScopeNode`, and their `hasFocus` returns true, then after calling `unfocus()`, it will return false.  Before this PR, it would only do that if `hasPrimaryFocus` was also true.

This also fixes a bug in the way `setFirstFocus` was implemented, making it conform more to the behavior of the previous implementation.  It has simplified logic in `reparent`, and in when it requests focus for scope nodes that have had `setFirstFocus` called on them.

## Related Issues

Fixes #31784 
Fixes #7247
Fixes #31827

## Tests

I added the following tests:
 - A test for unfocus.
 - Test that setting first focus requests focus for the scope properly.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 1]
```diff
@@ -61,6 +61,9 @@ class FocusAttachment {
   void detach() {
     assert(_node != null);
     if (isAttached) {
+      if (_node.hasPrimaryFocus) {
+       _node.unfocus();
```
Comment (darrenaustin) [inline packages/flutter/lib/src/widgets/focus_manager.dart]: Minor nit, missing a space of indentation?

**Review by darrenaustin — APPROVED** [review]
Again, not familiar with this code, but it LGTM.

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/focus_manager.dart]: Oooh, good eye.  Yes, I'll fix that.
